api/src/main/java/com/datatorrent/api/Launcher.java

```diff
@@ -0,0 +1,111 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package com.datatorrent.api;
+
+import java.util.Iterator;
+import java.util.ServiceLoader;
+
+import org.apache.hadoop.conf.Configuration;
+
+/**
+ *
+ */
+public abstract class Launcher<H extends Launcher.AppHandle>
+{
+
+  public enum LaunchMode
+  {
+    LOCAL, CLUSTER
+  }
+
+  public interface AppHandle {}
+
+  /**
+   * Get the launcher instance
+   *
+   * @param launchMode - The launch mode to use
+   *
+   * @return The launcher
+   */
+  public static Launcher<?> getLauncher(LaunchMode launchMode)
+  {
+    Launcher launcher;
+    if (launchMode == LaunchMode.LOCAL) {
+      launcher = LocalMode.newInstance();
+    } else {
+      launcher = ClusterMode.newInstance();
+    }
+    return launcher;
+  }
+
+  /**
+   * Launch application with configuration
+   *
+   * @param application  - Application to be run
+   * @param configuration - Application Configuration
+   *
+   * @return The application handle
+   */
+  public H launchApp(StreamingApplication application, Configuration configuration) throws LaunchException
+  {
+    return launchApp(application, configuration, null);
+  }
+
+  /**
+   * Launch application with configuration and launch parameters
+   *
+   * @param application  - Application to be run
+   * @param configuration - Application Configuration
+   * @param launchAttributes - Launch Parameters
+   *
+   * @return The application handle
+   */
+  public abstract H launchApp(StreamingApplication application, Configuration configuration, Attribute.AttributeMap launchAttributes) throws LaunchException;
+
+  /**
+   * Shutdown the application
+   *
+   * @param app
+   */
+  public abstract void shutdownApp(H app);
+
+  protected static <T> T loadService(Class<T> clazz)
+  {
+    ServiceLoader<T> loader = ServiceLoader.load(clazz);
+    Iterator<T> impl = loader.iterator();
+    if (!impl.hasNext()) {
+      throw new RuntimeException("No implementation for " + clazz);
+    }
+    return impl.next();
+  }
+
+  public static class LaunchException extends Exception
```

Comment:
You can have your own unchecked exception if you think there are cases where the user may want to handle them different from other errors. It does not require checked exceptions that are just getting into the way when the caller is not interested.